spec/environment_spec.rb

```diff
@@ -551,4 +551,268 @@ def new_content(id)
+    x509 = OpenSSL::X509::Certificate.new(ent['certificates'][0]['cert'])
+    extensions_hash = Hash[x509.extensions.collect { |ext| [ext.oid, ext.value] }]
+
+    expect(extensions_hash.has_key?("1.3.6.1.4.1.2312.9.2.#{contentB['id']}.1")).to be true
+
+    consumer_cp.update_consumer({:environments => [{:name => @env.name}, {:name => envC.name}, {:name => envB.name}]})
+    ent = consumer_cp.list_entitlements()[0]
+    x509 = OpenSSL::X509::Certificate.new(ent['certificates'][0]['cert'])
+    extensions_hash = Hash[x509.extensions.collect { |ext| [ext.oid, ext.value] }]
+
+    expect(extensions_hash.has_key?("1.3.6.1.4.1.2312.9.2.#{contentB['id']}.1")).to be true
+  end
+
+  it 'should NOT re-gen consumer entitlement cert when environment priority are reordered and entitlement does not provide specific content' do
+    consumer = @org_admin.register(random_string('testsystem'), :system, nil, {}, nil, nil, [], [],
+      [{ 'id' => @env['id']}])
+
+    expect(consumer['environments']).not_to be_nil
+
+    consumer_cp = Candlepin.new(nil, nil, consumer['idCert']['cert'], consumer['idCert']['key'])
+    product = create_product
+    contentA = create_content
+    contentB = create_content
+    contentC = create_content
+    contentD = create_content # not added to product
+    @cp.add_content_to_product(@owner['key'], product['id'], contentA['id'])
+    @cp.add_content_to_product(@owner['key'], product['id'], contentB['id'])
+    @cp.add_content_to_product(@owner['key'], product['id'], contentC['id'])
+    envB = @org_admin.create_environment(@owner['key'], 'envB', "envB", "For test systems only.")
+    consumer_cp.update_consumer({:environments => [{:name => @env.name}, {:name => envB.name}]})
+    job = @org_admin.promote_content(@env['id'],[{:contentId => contentB['id'],}])
+    wait_for_job(job['id'], 15)
+    job = @org_admin.promote_content(envB['id'],[{:contentId => contentD['id'],}])
+    wait_for_job(job['id'], 15)
+    pool = create_pool_and_subscription(@owner['key'], product['id'], 10)
+    ent = consumer_cp.consume_pool(pool['id'], {:quantity => 1})[0]
+    x509 = OpenSSL::X509::Certificate.new(ent['certificates'][0]['cert'])
+    extensions_hash = Hash[x509.extensions.collect { |ext| [ext.oid, ext.value] }]
+
+    expect(extensions_hash.has_key?("1.3.6.1.4.1.2312.9.2.#{contentB['id']}.1")).to be true
+    expect(extensions_hash.has_key?("1.3.6.1.4.1.2312.9.2.#{contentD['id']}.1")).to be false
+
+    consumer_cp.update_consumer({:environments => [{:name => envB.name}, {:name => @env.name}]})
+    ent = consumer_cp.list_entitlements()[0]
+    x509 = OpenSSL::X509::Certificate.new(ent['certificates'][0]['cert'])
+    extensions_hash = Hash[x509.extensions.collect { |ext| [ext.oid, ext.value] }]
+
+    expect(extensions_hash.has_key?("1.3.6.1.4.1.2312.9.2.#{contentB['id']}.1")).to be true
+    expect(extensions_hash.has_key?("1.3.6.1.4.1.2312.9.2.#{contentD['id']}.1")).to be false
+  end
```

Comment:
Let's also add spec tests for the removal scenarios

Reply:
Done